rust/wasmtime_bindings/src/memory_creator.rs

```diff
@@ -0,0 +1,95 @@
+use anyhow::{anyhow, Result};
+use std::{cmp, ptr, slice, u32};
+use wasmtime::LinearMemory;
+
+const WASM_PAGE_SIZE: usize = 64 * 1024;
+
+extern "C" {
+    fn aligned_alloc(align: usize, size: usize) -> *mut u8;
+    fn free(ptr: *mut u8);
+}
+
+pub struct ScyllaLinearMemory {
+    ptr: *mut u8,
+    size: usize,
+    maximum_size: Option<usize>,
+}
+
+// The entire ScyllaLinearMemory is used only in a single thread,
+// because we're not sharing it between seastar shards
+unsafe impl Send for ScyllaLinearMemory {}
+unsafe impl Sync for ScyllaLinearMemory {}
+
+impl Drop for ScyllaLinearMemory {
+    fn drop(&mut self) {
+        // previously allocated or reset to nullptr in grow_to()
+        unsafe { free(self.ptr) };
+    }
+}
+
+unsafe impl LinearMemory for ScyllaLinearMemory {
+    fn byte_size(&self) -> usize {
+        self.size
+    }
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.maximum_size
+    }
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        if new_size == self.size {
+            return Ok(());
+        }
+        let max_size = self.maximum_size.unwrap_or(u32::MAX as usize);
+        assert!(new_size <= max_size);
+        let new_ptr: *mut u8;
+        if new_size == 0 {
+            new_ptr = ptr::null_mut()
+        } else {
+            new_ptr = unsafe { aligned_alloc(WASM_PAGE_SIZE, new_size) };
```

Comment:
This answers my previous question.

aligned_alloc isn't defined if new_size isn't a multiple of WASM_PAGE_SIZE, is it guaranteed by the caller?

Reply:
When I look at our `aligned_alloc` overwrite it looks like we align the size up when necessary, but I see that in the linux specification it's as you described.
I'd expect the size here to be aligned, but I don't see it specified directly, so I think I'll just use memalign here

Reply:
(you could just align the new size yourself)

Reply:
I'll align the size myself then